packages/core/src/extensions/Comments/CommentsPlugin.ts

```diff
@@ -0,0 +1,263 @@
+import { User } from "../../models/User.js";
+import { EventEmitter } from "../../util/EventEmitter.js";
+import { ThreadStore } from "./threadstore/ThreadStore.js";
+import { CommentBody, ThreadData } from "./types.js";
+import { UserStore } from "./userstore/UserStore.js";
+const PLUGIN_KEY = new PluginKey(`blocknote-comments`);
+
+enum CommentsPluginActions {
+  SET_SELECTED_THREAD_ID = "SET_SELECTED_THREAD_ID",
+}
+
+type CommentsPluginAction = {
+  name: CommentsPluginActions;
+  data: string | null;
+};
+
+type CommentsPluginState = {
+  threadPositions: Map<string, { from: number; to: number }>;
+  // selectedThreadId: string | null;
+  // selectedThreadPos: number | null;
+  decorations: DecorationSet;
+};
+
+function updateState(
+  doc: Node,
+  selectedThreadId: string | undefined,
+  markType: string
+) {
+  const threadPositions = new Map<string, { from: number; to: number }>();
+  const decorations: Decoration[] = [];
+  // find all thread marks and store their position + create decoration for selected thread
+  doc.descendants((node, pos) => {
+    node.marks.forEach((mark) => {
+      if (mark.type.name === markType) {
+        const thisThreadId = (mark.attrs as { threadId: string | undefined })
+          .threadId;
+        if (!thisThreadId) {
+          return;
+        }
+        const from = pos;
+        const to = from + node.nodeSize;
+
+        // FloatingThreads component uses "to" as the position, so always store the largest "to" found
+        // AnchoredThreads component uses "from" as the position, so always store the smallest "from" found
+        const currentPosition = threadPositions.get(thisThreadId) ?? {
+          from: Infinity,
+          to: 0,
+        };
+        threadPositions.set(thisThreadId, {
+          from: Math.min(from, currentPosition.from),
+          to: Math.max(to, currentPosition.to),
+        });
+
+        if (selectedThreadId === thisThreadId) {
+          decorations.push(
+            Decoration.inline(from, to, {
+              class: "bn-thread-mark-selected",
+            })
+          );
+        }
+      }
+    });
+  });
+  return {
+    decorations: DecorationSet.create(doc, decorations),
+    threadPositions,
+  };
+}
+
+export class CommentsPlugin extends EventEmitter<any> {
+  public readonly plugin: Plugin;
+  public readonly userStore: UserStore<User>;
+
+  private pendingComment = false;
+  private selectedThreadId: string | undefined;
+
+  private emitStateUpdate() {
+    this.emit("update", {
+      selectedThreadId: this.selectedThreadId,
+      pendingComment: this.pendingComment,
+    });
+  }
+
+  /**
+   * when a thread is resolved or deleted, we need to update the marks to reflect the new state
+   */
+  private updateMarksFromThreads = (threads: Map<string, ThreadData>) => {
+    const doc = new Y.Doc();
+    const threadMap = doc.getMap("threads");
+    threads.forEach((thread) => {
+      threadMap.set(thread.id, thread);
+    });
```

Comment:
Is this doing anything? I might be missing something here

Reply:
lol thx